File: 20230109_Monday/20230109.go
// Leetcode: 144. Binary Tree Preorder Traversal.
// https://leetcode.com/problems/binary-tree-preorder-traversal/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 2 ms, faster than 34.10% of Go online submissions for Binary Tree Preorder Traversal.
// Memory Usage: 1.9 MB, less than 65.48% of Go online submissions for Binary Tree Preorder Traversal.
// 2023.01.09 Daily Challenge.
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func preorderTraversal(root *TreeNode) []int {
  answer := make([]int,0)
  if nil == root { return answer }
  q := make([]*TreeNode,0)
  q = append(q,root)
  for 0 != len(q){
    x := q[len(q)-1]
    q = q[0:len(q)-1]
    answer = append(answer,x.Val)
    if nil != x.Right { q = append(q, x.Right) }
    if nil != x.Left { q = append(q, x.Left) }
  }
  return answer
}

File: 20230109_Monday/20230109.rb
# Leetcode: 144. Binary Tree Preorder Traversal.
# https://leetcode.com/problems/binary-tree-preorder-traversal/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 87 ms, faster than 75.68% of Ruby online submissions for Binary Tree Preorder Traversal.
# Memory Usage: 210.8 MB, less than 100.00% of Ruby online submissions for Binary Tree Preorder Traversal.
# 2023.01.09 Daily Challenge.
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer[]}
def preorder_traversal(root)
  return [] if root.nil?
  q = [root]
  answer = []
  while ! q.empty?
    x = q.pop()
    answer.push(x.val)
    q.push(x.right) if x.right
    q.push(x.left) if x.left
  end
  return answer
end
